ocaml/xcp-rrdd/bin/rrdd/rrdd_monitor.ml

```diff
@@ -62,14 +62,58 @@ let merge_new_dss rrd dss =
     )
     rrd new_dss
 
+module OwnerMap = Map.Make (struct
+  type t = ds_owner
+
+  let compare a b =
+    match (a, b) with
+    | Host, Host ->
+        0
+    | Host, _ | VM _, SR _ ->
+        -1
+    | _, Host | SR _, VM _ ->
+        1
+    | VM a, VM b | SR a, SR b ->
+        String.compare a b
+end)
+
+let owner_to_string () = function
+  | Host ->
+      "host"
+  | VM uuid ->
+      "VM " ^ uuid
+  | SR uuid ->
+      "SR " ^ uuid
+
 (** Updates all of the hosts rrds. We are passed a list of uuids that is used as
     the primary source for which VMs are resident on us. When a new uuid turns
     up that we haven't got an RRD for in our hashtbl, we create a new one. When
     a uuid for which we have an RRD for doesn't appear to have any stats this
     update, we assume that the domain has gone and we stream the RRD to the
     master. We also have a list of the currently rebooting VMs to ensure we
     don't accidentally archive the RRD. *)
-let update_rrds timestamp dss (uuid_domids : (string * int) list) paused_vms =
+let update_rrds timestamp dss uuid_domids paused_vms =
+  let uuid_domids = List.to_seq uuid_domids |> StringMap.of_seq in
+  let paused_vms = List.to_seq paused_vms |> StringSet.of_seq in
+  let consolidate all (owner, ds) =
+    let add_ds_to = StringMap.add ds.ds_name ds in
+    let merge = function
+      | None ->
+          Some (add_ds_to StringMap.empty)
+      | Some dss ->
+          warn {|%s: found duplicate datasource with key "%s" in owner "%a"|}
+            __FUNCTION__ ds.ds_name owner_to_string owner ;
+          Some (add_ds_to dss)
+    in
+    OwnerMap.update owner merge all
+  in
+  let dss = List.to_seq dss |> Seq.fold_left consolidate OwnerMap.empty in
```

Comment:
This could've been just a regular fold_left, avoids constructing a seq just to immediately fold on it (although if you intend to push the list -> seq to the caller might be worth keeping)